Avoid UI functions in non-main thread in wxGStreamerMediaBackend

swt2c opened the pull request:
In aa2cd42, a couple of calls to NotifyMovieSizeChanged() were added, presumably for a good reason.  Unfortunately, NotifySizeChanged() makes UI calls on wxGTK, so this resulted in UI calls happening on a non-main thread because many of GStreamer's callbacks occur on arbitrary threads.  This is bad because it caused random X11 crashes.

Fix this by defining a new internal event type and queuing an event so that the main thread can call NotifyMovieSizeChanged().

vadz commented:
Thanks for fixing this, but I wonder if we could just use `CallAfter()` here instead of adding another ad hoc event?

I could modify the code to do it, if you're not sure about using it, but I can't test this myself because I don't know how to reproduce the problem.

swt2c commented:
> Thanks for fixing this, but I wonder if we could just use CallAfter() here instead of adding another ad hoc event?

Shoot - I swear I grepped the code for CallAfter() and couldn't find it - and then I convinced myself that it was something introduced in wxPython.  That's a much better solution.

> I could modify the code to do it, if you're not sure about using it, but I can't test this myself because I don't know how to reproduce the problem.

Weird, I see it while basically trying to do anything the the mediactrl sample.  It's strange, though, that no one has reported this before as the problem's theoretically been there for 4+ years.  :-)  I'll switch to CallAfter and retest.

swt2c committed:
Avoid UI functions in non-main thread in wxGStreamerMediaBackend

In aa2cd42, a couple of calls to NotifyMovieSizeChanged() were added,
presumably for a good reason.  Unfortunately, NotifySizeChanged() makes
UI calls on wxGTK, so this resulted in UI calls happening on a non-main
thread because many of GStreamer's callbacks occur on arbitrary threads.
This is bad because it caused random X11 crashes.

Fix this by calling NotifyMovieSizeChanged() with CallAfter().

swt2c commented:
Okay, here's hopefully a slightly better implementation.  Unfortunately it was not as straightforward as one would hope because `wxGStreamerMediaBackend` is not a `wxEvtHandler` and thus could not use `CallAfter()` directly.  I had to use its friend, `wxGStreamerMediaEventHandler`, as a proxy.

vadz commented:
I think we could derive `wxGStreamerMediaBackend` from `wxEvtHandler` if we really wanted, but as we already have `wxGStreamerMediaEventHandler` I don't see anything wrong in reusing it for this, so I'll just merge this version. Thanks again!
